### ISSUE TITLE
[functorch] AOT module does not run custom backwards pass compiler

### ISSUE THREAD

**Issue by lessw2020** [pytorch/pytorch#86373]
### 🐛 Describe the bug

Using nightly 1.13.0.dev20220928+cu116.

Pass in custom fwd and bwd compiler to aot_module:
~~~
self.compiled_m = aot_module(
                mod=self.module, 
                fw_compiler=self._compile_fwd, 
                bw_compiler=self._compile_bwd,
                partition_fn = default_partition
            )
~~~
will result in the fwd compiler being utilized, but the bwd_compiler is never invoked.
Debugging into AOT, I find that the aot config is properly set with both forward and backward, and a forward and backward graph and module are created in create_joint_forward_backward.

However, only the fwd compiler is called and not the backward compiler.

I found I can get the backward compiler called by making the following addition to aot_dispatch_autograd just after the track_graph_compiling for fwd:
~~~
with track_graph_compiling("backward"):
            compiled_bw_func = aot_config.bw_compiler(bw_module, deduped_flat_args)
~~~

At which point, both fwd and bwd compilers are called as expected and I can run full and successful iterations.

In reviewing the previous PR's for aot_autograd, it seems that area of the code was modified with this PR:
https://github.com/pytorch/pytorch/commit/5e8b4c64aa1ecd2e56aeeb06559af97162effdd1

and potentially the bwd compiler call was lost there.

Regardless, would like to ensure that bwd compiler can be invoked and thus this issue.  Not sure if my fix above is the proper/full fix, but gets us unblocked for now. 

### Versions

PyTorch version: 1.13.0.dev20220928+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.3
Libc version: glibc-2.31

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1020-aws-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration: 
GPU 0: NVIDIA A10G
GPU 1: NVIDIA A10G
GPU 2: NVIDIA A10G
GPU 3: NVIDIA A10G

Nvidia driver version: 510.73.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.13.0.dev20220928+cu116
[pip3] torch-model-archiver==0.5.3b20220226
[pip3] torch-workflow-archiver==0.2.4b20220513
[pip3] torchserve==0.6.0b20220513
[pip3] torchtext==0.13.1
[pip3] torchvision==0.14.0.dev20220928+cu116
[pip3] vit-pytorch==0.35.8
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0            16_linux64_mkl    conda-forge
[conda] captum                    0.5.0                         0    pytorch
[conda] cudatoolkit               11.6.0              hecad31d_10    conda-forge
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            16_linux64_mkl    conda-forge
[conda] magma-cuda116             2.6.1                         1    pytorch
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.23.3           py39hba7629e_0    conda-forge
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.13.0.dev20220928+cu116          pypi_0    pypi
[conda] torch-model-archiver      0.5.3                    py39_0    pytorch
[conda] torch-workflow-archiver   0.2.4                    py39_0    pytorch
[conda] torchserve                0.6.0                    py39_0    pytorch
[conda] torchtext                 0.13.1                     py39    pytorch
[conda] torchvision               0.14.0.dev20220928+cu116          pypi_0    pypi
[conda] vit-pytorch               0.35.8                   pypi_0    pypi

cc @zou3519 @Chillee @samdow @soumith

**Comment by bdhirsh** [comment]
The bw_compiler() call is here: https://github.com/pytorch/pytorch/blob/08780229df8f860dbef3fa82ffd1072b124b29c5/functorch/_src/aot_autograd.py#L422

One guess I have: do any of the input tensors that you trace with have `requires_grad=True`? Today, if no tensor inputs require gradients, then `aot_module()` won't bother to trace a backward graph at all, since there's no need for autograd: https://github.com/pytorch/pytorch/blob/08780229df8f860dbef3fa82ffd1072b124b29c5/functorch/_src/aot_autograd.py#L507

**Comment by lessw2020** [comment]
Hi @bdhirsh,
Thanks for the comments.  I did already see the 422 line you noted, the issue though is that line is never hit when debugging, thus I moved up and added in the tracking call to force it to fire.

Re: requires_grad - great idea to check and good info to know.   
For the current issue, I confirm that we are passing in tensors with requires_grad= True:
<img width="1058" alt="requires_grad_true" src="https://user-images.githubusercontent.com/46302957/194415426-d69636c8-ef48-4b31-95ce-bcb97991183a.png">

Specifically right now we're making a baby model with nn.Linear layers stacked in a nn.Sequential. But the parameters are all marked with requires_grad, which I verified.

Also, if it's helpful, you can repro this using this file/repo:
https://github.com/lessw2020/comm_fusion/blob/main/demo/fsdp_master.py

you can git clone and just run (within the demo dir): python fsdp_master.py. 
You'll then see that the backwards compiler (Engine._compile_bwd) is never called unless you add the call I noted above (and made in the file aot_autograd_fixed.py, in the same dir). 

Thanks again for the info above.

**Comment by lessw2020** [comment]
Just to expand on the status of requires_grad - we are all_gathering params, so I made sure to also check that the all_gathered params (and not just the starting model param) is also marked with requires grad:

<img width="432" alt="all_gather_grad" src="https://user-images.githubusercontent.com/46302957/194426266-451aa7ff-0cd3-4017-b647-3874726144af.png">


<img width="774" alt="sharded_param_grad" src="https://user-images.githubusercontent.com/46302957/194424918-58bc31ab-cbce-4980-bc83-1c2b7b98f1e5.png">

And also put some monitors for the needs_autograd flag within AOT to make sure everything is as expected b/c we are doing the all_gather within a no grad context.

<img width="774" alt="aot_autograd_status" src="https://user-images.githubusercontent.com/46302957/194425012-ea39e4d7-d47d-401b-b7cf-3c45032788b4.png">

And without the track_call above, we still do no have our bwd_ compile called.

**Comment by lessw2020** [comment]
update - even easier repo. 
I tested the tutorial that is found here:
https://pytorch.org/functorch/stable/notebooks/aot_autograd_optimizations.html

As you can see, in the original tutorial code, both the fwd and backward compiler are called.  
Screenshot (note 2nd print with tangents)
<img width="1142" alt="tutorial_bwd_pass" src="https://user-images.githubusercontent.com/46302957/194782687-097dc838-db35-4832-8837-2d4414a3dc0b.png">

By contrast, running the tutorial with latest nightlies = no bwd pass. 

<img width="879" alt="current_tutorial_no_bwd" src="https://user-images.githubusercontent.com/46302957/194782745-0f5e6b52-0c67-49c1-bc0b-fd04a8dc8872.png">


I hope that substantially helps reduce the repo case to get this resolved.

**Comment by Chillee** [comment]
@lessw2020 Now, we no longer do compilation until you actually run the backwards pass. So you need to do `aot_print_fn` and then call `.backward()`

**Comment by lessw2020** [comment]
@Chillee  - thanks for the update.  

Can you advise though - I'm not clear then how we would modify the bwd graph before executing, if we have to actually execute .backward() in the first place to get the graph? 
(i.e. we don't run the forward pass to get the fwd graph, we get it as a result of calling aot_module and apparently used to get both graphs).   
Greatly appreciate any insight here.
